framework/src/meshgenerators/MeshGenerator.C

```diff
@@ -20,15 +20,58 @@ MeshGenerator::validParams()
                         false,
                         "Whether or not to show mesh info after generating the mesh "
                         "(bounding box, element types, sidesets, nodesets, subdomains, etc)");
+  params.addParam<std::vector<std::string>>(
+      "selected_mesh_metadata_to_retain",
+      std::vector<std::string>(),
+      "Names of input mesh's metadata that are selected to retain.");
+  params.addParam<bool>("retain_all_input_mesh_metadata",
+                        false,
+                        "Whether to retain all the mesh metadata of the input mesh.");
 
   params.registerBase("MeshGenerator");
 
   return params;
 }
 
 MeshGenerator::MeshGenerator(const InputParameters & parameters)
-  : MooseObject(parameters), MeshMetaDataInterface(this), _mesh(_app.actionWarehouse().mesh())
+  : MooseObject(parameters),
+    MeshMetaDataInterface(this),
+    _mesh(_app.actionWarehouse().mesh()),
+    _selected_mesh_metadata_to_retain(
+        getParam<std::vector<std::string>>("selected_mesh_metadata_to_retain")),
+    _retain_all_input_mesh_metadata(getParam<bool>("retain_all_input_mesh_metadata"))
 {
+  if (isParamValid("input"))
+  {
+    const MeshGeneratorName input_name(getParam<MeshGeneratorName>("input"));
+    if (_retain_all_input_mesh_metadata)
+    {
+      if (!_selected_mesh_metadata_to_retain.empty())
+        paramError(
+            "selected_mesh_metadata_to_retain",
+            "This parameter should not be provided if retain_all_input_mesh_metadata is set true.");
+      const auto mesh_metadata_names = findMeshMetaData(input_name);
+      for (const auto & mmd_name : mesh_metadata_names)
+        AddMeshMetaDataAlias(input_name, mmd_name, name(), mmd_name);
+    }
+    else
+      for (const auto & mmd_name : _selected_mesh_metadata_to_retain)
+      {
+        if (!hasMeshProperty(mmd_name, input_name))
+          paramError("selected_mesh_metadata_to_retain",
+                     "The specified mesh metadata to retain does not exist in the input mesh.");
+        AddMeshMetaDataAlias(input_name, mmd_name, name(), mmd_name);
+      }
+  }
```

Comment:
we should move some or most of this code inside the interface.
It s cleaner to have small minimal base classes